packages/ckeditor5-watchdog/src/actionsrecorder.ts

```diff
@@ -0,0 +1,780 @@
+
+	/**
+	 * Stack tracking nested action frames to maintain call hierarchy.
+	 */
+	private _frameStack: Array<ActionsRecorderEntry> = [];
+
+	/**
+	 * Set of already reported errors used to notify only once for each error (not on every try-catch nested block).
+	 */
+	private _errors = new Set<Error>();
+
+	/**
+	 * Maximum number of action entries to keep in memory.
+	 */
+	private _maxEntries: number;
+
+	/**
+	 * Error callback.
+	 */
+	private _errorCallback?: ActionsRecorderErrorCallback;
+
+	/**
+	 * Filter function to determine which entries should be stored.
+	 */
+	private _filterCallback?: ActionsRecorderFilterCallback;
+
+	/**
+	 * Callback triggered every time count of recorded entries reaches maxEntries.
+	 */
+	private _maxEntriesCallback: ActionsRecorderMaxEntriesCallback;
+
+	/**
+	 * @inheritDoc
+	 */
+	public static get pluginName() {
+		return 'ActionsRecorder' as const;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public static override get isOfficialPlugin(): true {
+		return true;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public constructor( editor: Editor ) {
+		super( editor );
+
+		editor.config.define( 'actionsRecorder.maxEntries', 1000 );
+
+		const config = editor.config.get( 'actionsRecorder' )!;
+
+		this._maxEntries = config.maxEntries!;
+		this._filterCallback = config.onFilter;
+		this._errorCallback = config.onError;
+		this._maxEntriesCallback = config.onMaxEntries || this._maxEntriesDefaultHandler;
+
+		this._tapCommands();
+		this._tapOperationApply();
+		this._tapModelMethods();
+		this._tapModelSelection();
+		this._tapComponentFactory();
+		this._tapViewDocumentEvents();
+	}
+
+	/**
+	 * Returns all recorded action entries.
+	 */
+	public getEntries(): Array<ActionsRecorderEntry> {
+		// Return a shallow copy instead of reference as this array could be modified.
+		return this._entries.slice();
+	}
+
+	/**
+	 * Flushes all recorded entries.
+	 */
+	public flushEntries(): void {
+		this._entries = [];
+	}
+
+	/**
+	 * Creates a new action frame and adds it to the recording stack.
+	 *
+	 * @param action The name/type of the action being recorded.
+	 * @param params Optional parameters associated with the event.
+	 * @returns The created call frame object.
+	 */
+	private _enterFrame( action: string, params?: Array<unknown> ): ActionsRecorderEntry {
+		const callFrame: ActionsRecorderEntry = {
+			timeStamp: new Date().toISOString(),
+			...this._frameStack.length && { parentEntry: this._frameStack.at( -1 ) },
+			action,
+			params: params?.map( param => serializeValue( param ) ),
+			before: this._buildStateSnapshot()
+		};
+
+		// Apply filter if configured, only add to entries if filter passes.
+		if ( !this._filterCallback || this._filterCallback( callFrame, this._entries ) ) {
+			// Add the call frame to the entries.
+			this._entries.push( callFrame );
+
+			// Enforce max entries limit.
+			if ( this._entries.length >= this._maxEntries ) {
+				this._maxEntriesCallback();
```

Comment:
Calling `_maxEntriesCallback()` in `_enterFrame` instead of `_leaveFrame` (or, alternatively, adding it to `_entries` here) means that the last entry will have no `after`, `result` nor `error` fields properly set. If the integrator follows the examples in the docs for `onMaxEntries`, they'll call `flushEntries()` and this record would be already lost when `after`, `result` or `error` is available.

Reply:
Actually, the problem is more in adding records to `_entries` than calling `_maxEntriesCallback`. Consider the scenario when the integrator sets `maxEntries` to infinity and checks them in `setInterval` - they could read and clear incomplete frames.